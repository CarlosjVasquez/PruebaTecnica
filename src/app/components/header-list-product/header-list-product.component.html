<div class="header">
  <search (onInputEventemitter)="onHandleSubmit($event)" />
  <custom-button [label]="'Agregar'" (onEventClick)="onAddProduct()" />
</div>
