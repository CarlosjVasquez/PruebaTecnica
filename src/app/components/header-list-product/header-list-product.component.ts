import { Component, EventEmitter, Output } from '@angular/core';
import { ButtonComponent } from '../button/button.component';
import { SearchComponent } from '../search/search.component';
import { Router } from '@angular/router';

@Component({
  selector: 'header-list-product',
  standalone: true,
  imports: [ButtonComponent, SearchComponent],
  templateUrl: './header-list-product.component.html',
  styleUrl: './header-list-product.component.scss',
})
export class HeaderListProductComponent {
  @Output() onInputEventemitter: EventEmitter<string> = new EventEmitter();
  constructor(private router: Router) {}

  onAddProduct() {
    this.router.navigate(['/product/create']);
  }

  onHandleSubmit(search: string) {
    this.onInputEventemitter.emit(search);
  }
}
