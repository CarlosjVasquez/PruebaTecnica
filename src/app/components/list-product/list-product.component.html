<div class="container">
  <header-list-product (onInputEventemitter)="onSearch($event)" />
  <div class="my-3 p-4 bg-white">
    <div class="content-table">
      <custom-table
        class="table"
        [data]="filterListData"
        [columns]="listColumns"
        [options]="options"
        (onDropdownClickEventEmitter)="onControlEvents($event)"
      ></custom-table>
    </div>
    <div class="footer-table mt-3">
      <results [results]="listData.length || 0" label="Resultados" />
      <custom-select
        [options]="optionsItemPage"
        (onChangeEventEmmiter)="onHandleSelectItemPage($event)"
      />
    </div>
  </div>
</div>

<custom-modal
  [show]="showModal"
  (onCancelEventEmmiter)="cancelModal()"
  (onSuccessEventEmmiter)="confirmDelete()"
  description="¿Estás seguro de eliminar el produto {{ selectedProduct.name }}?"
/>
