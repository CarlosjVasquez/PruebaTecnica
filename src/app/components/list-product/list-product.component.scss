.content-table {
  width: 100%;
  overflow-y: auto;
}

.footer-table {
  width: 100%;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}
