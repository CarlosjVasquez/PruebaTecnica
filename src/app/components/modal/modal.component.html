@if (show) {
<div class="modal">
  <div class="modal-content">
    <div class="modal-body">
      <p>{{ description }}</p>
    </div>

    <div class="modal-footer pt-3">
      <div class="row">
        <div class="col">
          <custom-button
            (onEventClick)="onCancel()"
            class="custom-btn"
            color="body"
            label="Cancelar"
          />
        </div>
        <div class="col">
          <custom-button
            (onEventClick)="onSuccess()"
            class="custom-btn"
            label="Confirmar"
          />
        </div>
      </div>
    </div>
  </div>
</div>
}
