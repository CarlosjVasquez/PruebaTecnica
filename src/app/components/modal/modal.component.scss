/* Estilos generales para el modal */
.modal {
  /* Ocultar el modal por defecto */
  position: fixed;
  z-index: 1000;
  left: 0;
  top: 0;
  width: 100vw;
  height: 100vh;
  background-color: rgba(0, 0, 0, 0.5); /* Fondo semi-transparente */
  overflow: auto;
  display: flex;
  align-items: center;
  justify-content: center;
}

/* Estilos para el contenido del modal */
.modal-content {
  background-color: #fefefe;
  padding: 20px;
  border: 1px solid #888;
  width: 80%;
  max-width: 500px;
  position: relative;
  border-radius: 10px;
  .modal-body {
    // min-height: 100px;
    text-align: center;
    p {
    }
  }
  .modal-footer {
    border-top: 1px #888 solid;
  }
}

/* Estilos para el bot√≥n de cerrar */
.close {
  color: #aaa;
  float: right;
  font-size: 28px;
  font-weight: bold;
  cursor: pointer;
}

.close:hover,
.close:focus {
  color: black;
  text-decoration: none;
  cursor: pointer;
}

.custom-btn {
  display: flex;
  align-items: center;
  justify-content: center;
}
