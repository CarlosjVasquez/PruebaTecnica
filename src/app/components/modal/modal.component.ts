import { Component, EventEmitter, Input, Output } from '@angular/core';
import { ButtonComponent } from '../button/button.component';

@Component({
  selector: 'custom-modal',
  standalone: true,
  imports: [ButtonComponent],
  templateUrl: './modal.component.html',
  styleUrl: './modal.component.scss',
})
export class ModalComponent {
  @Input() description: string = '';
  @Input() show: boolean = false;
  @Output() onCancelEventEmmiter: EventEmitter<void> = new EventEmitter();
  @Output() onSuccessEventEmmiter: EventEmitter<void> = new EventEmitter();

  onCancel() {
    this.onCancelEventEmmiter.emit();
  }
  onSuccess() {
    this.onSuccessEventEmmiter.emit();
  }
}
