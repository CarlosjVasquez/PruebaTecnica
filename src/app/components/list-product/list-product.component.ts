import { Component, OnDestroy, OnInit } from '@angular/core';
import { HeaderListProductComponent } from '../header-list-product/header-list-product.component';
import {
  Columns,
  Data,
  DropdownEvent,
  TableComponent,
} from '../table/table.component';
import { ResultsComponent } from '../results/results.component';
import { OtionsSelect, SelectComponent } from '../select/select.component';
import { ModalComponent } from '../modal/modal.component';
import { Product } from '../../models/product';
import { ProductService } from '../../services/product.service';
import { Option } from '../dropdown/dropdown.component';
import { Router } from '@angular/router';
import { Subscription } from 'rxjs';

@Component({
  selector: 'list-product',
  standalone: true,
  imports: [
    HeaderListProductComponent,
    TableComponent,
    ResultsComponent,
    SelectComponent,
    ModalComponent,
  ],
  templateUrl: './list-product.component.html',
  styleUrl: './list-product.component.scss',
})
export class ListProductComponent implements OnDestroy {
  private subscriptionProducts: Subscription;
  private subscriptionSelectedProduct: Subscription;
  listColumns: Columns[] = [
    {
      key: 'logo',
      label: 'Logo',
      type: 'image',
      align: 'center',
      width: 100,
    },
    {
      key: 'name',
      label: 'Nombre del Producto',
      type: 'text',
      align: 'center',
      width: 100,
    },
    {
      key: 'description',
      label: 'Descripción',
      type: 'text',
      align: 'center',
      width: 100,
      popover: 'Descripción corta del producto',
    },
    {
      key: 'date_release',
      label: 'Fecha de liberación',
      type: 'date',
      align: 'start',
      width: 100,
      popover: 'Fecha de liberación del producto',
    },
    {
      key: 'date_revision',
      label: 'Fecha de restructuración',
      type: 'date',
      align: 'start',
      width: 100,
      popover:
        'Fecha de restructuración del producto 1 año después de la fecha de liberación',
    },
  ];
  listData: Product[] = [];
  filterListData: Product[] = [];
  selectedProduct: Product = new Product('', '', '', '', '', '');
  optionsItemPage: OtionsSelect[] = [
    { label: '5', value: 5 },
    { label: '10', value: 10 },
    { label: '20', value: 20 },
  ];
  showModal: boolean = false;
  options: Option[] = [
    {
      id: 1,
      label: 'Eliminar',
    },
    {
      id: 2,
      label: 'Actualizar',
    },
  ];

  constructor(private router: Router, private productService: ProductService) {
    this.subscriptionProducts = this.productService.filterProducts$.subscribe(
      (value) => {
        this.listData = value;
        this.filterListData = value.slice(0, this.optionsItemPage[0].value);
      }
    );
    this.subscriptionSelectedProduct =
      this.productService.selectedProduct$.subscribe((value) => {
        this.selectedProduct = value;
      });
  }

  onHandleSelectItemPage(value: string | number) {
    this.filterListData = this.listData.slice(0, value as number);
  }

  onDeleteProduct(data: Data) {
    this.onHandleSelectProduct(data);
    this.openModal();
  }

  confirmDelete() {
    this.productService.deleteProduct();
    this.cancelModal();
  }

  onHandleSelectProduct(data: Data) {
    const product: Product = data as Product;
    this.productService.selectProduct(product);
  }

  onHandleUpdate(data: Data) {
    const productUpdate: Product = data as Product;
    this.onHandleSelectProduct(productUpdate);
    this.router.navigate(['/product/update']);
  }

  openModal() {
    this.showModal = true;
  }

  closeModal() {
    this.showModal = false;
  }

  cancelModal() {
    const product = new Product('', '', '', '', '', '');
    this.onHandleSelectProduct(product);
    this.closeModal();
  }

  onControlEvents({ id, data }: DropdownEvent) {
    if (id === 1) this.onDeleteProduct(data);
    if (id === 2) this.onHandleUpdate(data);
  }

  ngOnDestroy(): void {
    this.subscriptionProducts.unsubscribe();
    this.subscriptionSelectedProduct.unsubscribe();
  }

  onSearch(search: string) {
    this.productService.searchProduct(search);
  }
}
