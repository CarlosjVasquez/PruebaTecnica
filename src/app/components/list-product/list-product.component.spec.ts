import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ListProductComponent } from './list-product.component';
import { HeaderListProductComponent } from '../header-list-product/header-list-product.component';
import { TableComponent } from '../table/table.component';
import { ResultsComponent } from '../results/results.component';
import { SelectComponent } from '../select/select.component';
import { ModalComponent } from '../modal/modal.component';
import { of } from 'rxjs';
import { Product } from '../../models/product';
import { ProductService } from '../../services/product.service';
import { Router } from '@angular/router';

describe('ListProductComponent', () => {
  let component: ListProductComponent;
  let fixture: ComponentFixture<ListProductComponent>;
  let compiled: HTMLElement;
  let productServiceMock: any;
  let routerMock: any;

  beforeEach(async () => {
    productServiceMock = {
      filterProducts$: of([]),
      selectedProduct$: of(new Product('', '', '', '', '', '')),
      deleteProduct: jest.fn(),
      selectProduct: jest.fn(),
      searchProduct: jest.fn(),
    };

    routerMock = {
      navigate: jest.fn(),
    };

    await TestBed.configureTestingModule({
      providers: [
        ListProductComponent,
        HeaderListProductComponent,
        TableComponent,
        ResultsComponent,
        SelectComponent,
        ModalComponent,
        { provide: ProductService, useValue: productServiceMock },
        { provide: Router, useValue: routerMock },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ListProductComponent);
    component = fixture.componentInstance;
    compiled = fixture.nativeElement;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('match to snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should filter list data on item page selection', () => {
    component.listData = [
      new Product(
        '1',
        'Product 1',
        'Description 1',
        'logo1.png',
        '2023-01-01',
        '2024-01-01'
      ),
      new Product(
        '2',
        'Product 2',
        'Description 2',
        'logo2.png',
        '2023-01-02',
        '2024-01-02'
      ),
      new Product(
        '3',
        'Product 3',
        'Description 3',
        'logo3.png',
        '2023-01-03',
        '2024-01-03'
      ),
    ];
    component.onHandleSelectItemPage(2);
    expect(component.filterListData.length).toBe(2);
  });

  test('should open and close modal', () => {
    component.openModal();
    expect(component.showModal).toBeTruthy();

    component.closeModal();
    expect(component.showModal).toBeFalsy();
  });

  test('should confirm delete', () => {
    component.confirmDelete();
    expect(productServiceMock.deleteProduct).toHaveBeenCalled();
    expect(component.showModal).toBeFalsy();
  });

  test('should open and close modal', () => {
    component.openModal();
    expect(component.showModal).toBeTruthy();

    component.closeModal();
    expect(component.showModal).toBeFalsy();
  });

  test('should confirm delete', () => {
    component.confirmDelete();
    expect(productServiceMock.deleteProduct).toHaveBeenCalled();
    expect(component.showModal).toBeFalsy();
  });

  test('should handle select product', () => {
    const product = new Product(
      '1',
      'Product 1',
      'Description 1',
      'logo1.png',
      '2023-01-01',
      '2024-01-01'
    );
    component.onHandleSelectProduct(product);
    expect(productServiceMock.selectProduct).toHaveBeenCalledWith(product);
  });

  test('should handle update', () => {
    const product = new Product(
      '1',
      'Product 1',
      'Description 1',
      'logo1.png',
      '2023-01-01',
      '2024-01-01'
    );
    component.onHandleUpdate(product);
    expect(routerMock.navigate).toHaveBeenCalledWith(['/product/update']);
  });

  test('should handle control events', () => {
    const product = new Product(
      '1',
      'Product 1',
      'Description 1',
      'logo1.png',
      '2023-01-01',
      '2024-01-01'
    );

    component.onControlEvents({ id: 1, data: product });
    expect(component.showModal).toBeTruthy();

    component.onControlEvents({ id: 2, data: product });
    expect(routerMock.navigate).toHaveBeenCalledWith(['/product/update']);
  });

  test('should search product', () => {
    const search = 'Test';
    component.onSearch(search);
    expect(productServiceMock.searchProduct).toHaveBeenCalledWith(search);
  });

  test('should unsubscribe on destroy', () => {
    const subscriptionProductsSpy = jest.spyOn(
      component['subscriptionProducts'],
      'unsubscribe'
    );
    const subscriptionSelectedProductSpy = jest.spyOn(
      component['subscriptionSelectedProduct'],
      'unsubscribe'
    );

    component.ngOnDestroy();
    expect(subscriptionProductsSpy).toHaveBeenCalled();
    expect(subscriptionSelectedProductSpy).toHaveBeenCalled();
  });
});
