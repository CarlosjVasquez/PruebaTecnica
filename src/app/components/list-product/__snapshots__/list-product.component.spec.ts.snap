// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListProductComponent match to snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="container"
  >
    <header-list-product>
      <div
        class="header"
      >
        <search>
          <custom-input
            id="search"
            name="search"
            placeholder="Search..."
          >
            <div
              class="form-group"
            >
              
              <input
                class="form-input ng-untouched ng-pristine ng-valid"
                id="search"
                max="undefined"
                min="undefined"
                placeholder="Search..."
                type="text"
              />
              
            </div>
          </custom-input>
        </search>
        <custom-button>
          <button
            class="btn btn-md btn-primary"
            type="button"
          >
             Agregar

          </button>
        </custom-button>
      </div>
    </header-list-product>
    <div
      class="my-3 p-4 bg-white"
    >
      <div
        class="content-table"
      >
        <custom-table
          class="table"
        >
          <table
            class="table"
          >
            <thead>
              <tr>
                <th
                  style="min-width: 100px;"
                >
                  <div
                    class="content-column"
                    style="justify-content: center;"
                  >
                    <span
                      class="mr-2"
                    >
                      Logo
                    </span>
                    
                  </div>
                </th>
                <th
                  style="min-width: 100px;"
                >
                  <div
                    class="content-column"
                    style="justify-content: center;"
                  >
                    <span
                      class="mr-2"
                    >
                      Nombre del Producto
                    </span>
                    
                  </div>
                </th>
                <th
                  style="min-width: 100px;"
                >
                  <div
                    class="content-column"
                    style="justify-content: center;"
                  >
                    <span
                      class="mr-2"
                    >
                      Descripción
                    </span>
                    <popover
                      icon="assets/icons/informacion.png"
                    >
                      <div
                        class="content"
                      >
                        <custom-icon
                          class="popover-trigger"
                        >
                          <picture>
                            <img
                              alt="icon"
                              height="18"
                              src="assets/icons/informacion.png"
                              width="18"
                            />
                          </picture>
                        </custom-icon>
                        <div
                          class="popover-content"
                        >
                           Descripción corta del producto 
                        </div>
                      </div>
                    </popover>
                    
                  </div>
                </th>
                <th
                  style="min-width: 100px;"
                >
                  <div
                    class="content-column"
                    style="justify-content: start;"
                  >
                    <span
                      class="mr-2"
                    >
                      Fecha de liberación
                    </span>
                    <popover
                      icon="assets/icons/informacion.png"
                    >
                      <div
                        class="content"
                      >
                        <custom-icon
                          class="popover-trigger"
                        >
                          <picture>
                            <img
                              alt="icon"
                              height="18"
                              src="assets/icons/informacion.png"
                              width="18"
                            />
                          </picture>
                        </custom-icon>
                        <div
                          class="popover-content"
                        >
                           Fecha de liberación del producto 
                        </div>
                      </div>
                    </popover>
                    
                  </div>
                </th>
                <th
                  style="min-width: 100px;"
                >
                  <div
                    class="content-column"
                    style="justify-content: start;"
                  >
                    <span
                      class="mr-2"
                    >
                      Fecha de restructuración
                    </span>
                    <popover
                      icon="assets/icons/informacion.png"
                    >
                      <div
                        class="content"
                      >
                        <custom-icon
                          class="popover-trigger"
                        >
                          <picture>
                            <img
                              alt="icon"
                              height="18"
                              src="assets/icons/informacion.png"
                              width="18"
                            />
                          </picture>
                        </custom-icon>
                        <div
                          class="popover-content"
                        >
                           Fecha de restructuración del producto 1 año después de la fecha de liberación 
                        </div>
                      </div>
                    </popover>
                    
                  </div>
                </th>
                
                <th />
              </tr>
            </thead>
            <tbody>
              
              <tr>
                <td
                  class="not-data"
                  colspan="6"
                >
                  <span>
                    No hay datos registrados
                  </span>
                </td>
              </tr>
              
            </tbody>
          </table>
        </custom-table>
      </div>
      <div
        class="footer-table mt-3"
      >
        <results
          label="Resultados"
        >
          <div>
            0 Resultados
          </div>
        </results>
        <custom-select>
          <select
            class="ng-untouched ng-pristine ng-valid"
            id="itemPage"
            name="itemPage"
          >
            <option
              value="5"
            >
               5 
            </option>
            <option
              value="10"
            >
               10 
            </option>
            <option
              value="20"
            >
               20 
            </option>
            
          </select>
        </custom-select>
      </div>
    </div>
  </div>
  <custom-modal>
    
  </custom-modal>
</div>
`;
